### ISSUE TITLE
Refactoring of Kernel classes: MassTrace

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#1901]
I provide here some personal opinions and open for discussion:
In my view the current MassTrace implementation is too specialized, has inflexible design and is rather heave. It should be broken down in at least two classes.
1. A lightweight, minimal mass trace class which e.g. does not store both trace peaks and smoothed trace peaks, no additional statistics like centroid mz or sd, FWHM, quant method or label. Only a small set of members of general use should be provided (e.g. getIntensityWeightedCentroidMZ()) which compute information on the fly without storing them.
2. A specialized, heavy mass trace implementation which uses the lightweight class (e.g. via dependency injection) and e.g. may store computational expensive statistics etc like the current implementation.

**Comment by timosachsenberg** [comment]
also there should probably be two versions MassTrace1D (e.g. as base for simple XICs with single centroid MZ) and MassTrace2D (like now)

**Comment by timosachsenberg** [comment]
1D and 2D use case come up a few times in OpenMS and would def. make sense

**Comment by hroest** [comment]
:+1: many algorithms will create a lot of these objects so making them light is a good idea. 

btw: what is the difference between a MassTrace1D and a OpenMS::Chromatogram ? both describe 1D data in RT, right? However they are used differently, so I see the argument to keep them both. However, a Masstrace is basically used like a Feature, so why do we have both?

**Comment by timosachsenberg** [comment]
I think MassTrace1D could be the base for OpenMS::Chromatogram as it mainly adds meta data. For MassTrace2D I also think a lightweight version with Peak2D would be great. Additional functinality / meta values / statistics etc. could be part of derived classes.

**Comment by timosachsenberg** [comment]
To make this super consistent one should possibly also introduce a RawSpectrum1D and 2D class that only contains the raw peaks, the precursor annotation, the MS level and retention time without any additional meta information / float / string arrays etc. .

**Comment by hroest** [comment]
hmm I am not so sure what we should solve all our problems through inheritance ... 

I think it can also make sense to have well defined data structures for a certain type of algorithm and to have data structures that map 1:1 to the XML. I think one of the issues with OpenMS is that we tried to have both and now we are stuck with inefficient data structures but we are very good at reading and writing XML. So I think we should get away from finding a "one fits all solution" and *reduce* inheritance and complexity rather than increase it. We discussed this in Izmir, see all the good ideas

- https://github.com/OpenMS/OpenMS/wiki/OpenMS-2.0-Kernel-Whitepaper
- https://github.com/OpenMS/OpenMS/wiki/OpenMS-2.0-Planing-Interfaces-MSExperiment

and the sample code in #762 and  #763

**Comment by timosachsenberg** [comment]
I agree and think it is pretty much in line with the Izmir proposals.

**Comment by timosachsenberg** [comment]
Would be nice to revive the discussion

**Comment by hroest** [comment]
yes, I agree. However, more pressing would be probably to ensure (i) the quality of the current algorithms (e.g. that they produce state-of-the-art output compared to similar software) and (ii) the memory and runtime performance of these. For both of these tasks however we would need some data
